doc/man/doveadm-fts.1.in
.\" Copyright (c) 2014-2018 Dovecot authors, see the included COPYING file
.TH DOVEADM\-FTS 1 "2015-05-09" "Dovecot v2.3" "Dovecot"
.SH NAME
doveadm\-fts \- Manipulate the Full Text Search (FTS) index
.\"------------------------------------------------------------------------
.SH SYNOPSIS
.BR doveadm " [" \-Dv ]
.\"	[\fB\-f\fP \fIformatter\fP]
.BI fts \ command
.RI [ OPTIONS ]\ [ ARGUMENTS ]
.\"------------------------------------------------------------------------
.SH DESCRIPTION
The
doveadm fts
.I COMMANDS
can be used to manipulate the Full Text Search (FTS) index.
.\"------------------------------------------------------------------------
.\"	@INCLUDE:global-options-formatter@
@INCLUDE:global-options@
.\" --- command specific options --- "/.
.PP
This command uses by default the output formatter
.B flow
(without the
.IR key =
prefix).
.PP
Command specific
.IR options :
.\"-------------------------------------
@INCLUDE:option-A@
.\"-------------------------------------
@INCLUDE:option-F-file@
.\"-------------------------------------
@INCLUDE:option-S-socket@
.\"-------------------------------------
@INCLUDE:option-u-user@
.\"------------------------------------------------------------------------
.SH ARGUMENTS
.TP
.I namespace
The name of a namespace, e.g. the name of the shared namespace.
When no namespace was given, the user\(aqs private namespace will be used.
.\"------------------------------------------------------------------------
.SH COMMANDS
.SS fts optimize
.B doveadm fts optimize
[\fB\-u\fP \fIuser\fP|\fB\-A\fP|\fB\-F\fP \fIfile\fP]
[\fB\-S\fP \fIsocket_path\fP]
.RI [ namespace ]
.PP
Optimize the full text search index.
This is also done automatically by the full text search engines, but this
enforces it to happen.
.\"-------------------------------------
.SS fts rescan
.B doveadm fts rescan
[\fB\-u\fP \fIuser\fP|\fB\-A\fP|\fB\-F\fP \fIfile\fP]
[\fB\-S\fP \fIsocket_path\fP]
.RI [ namespace ]
.PP
Scan what mails exist in the full text search index and compare those to
what actually exist in mailboxes.
This removes mails from the index that have already been expunged and
makes sure that the next
.B doveadm index
will index all the missing mails (if any). Note that currently most FTS
backends do not implement this properly, but instead they delete all the
FTS indexes. This may change in the future versions.
.\"-------------------------------------
.SS fts check fast
.B doveadm fts check fast
[\fB\-u\fP \fIuser\fP|\fB\-A\fP|\fB\-F\fP \fIfile\fP]
[\fB\-S\fP \fIsocket_path\fP]
.RI [ \--refresh ]
.RI [ \--print-mismatches-only ]
.RI [ namespace ]
.PP
This command exists only when the fts_dovecot plugin (Dovecot Pro FTS) is
loaded.
.PP
This command can be used to check FTS indexes for consistency. It performs
a fast check using only information in local caches (fts.S, metacache).
.PP
.\"-------------------------------------
.TP
.B \--refresh
Refresh any necessary local caches for the command to run successfully.
This can be used if the command otherwise fails with exit code 68.
.\"-------------------------------------
.TP
.B \--print-mismatches-only
Print only mailboxes that have inconsistencies.
.\"-------------------------------------
.RE
.PP
Exit codes:
.PP
.\"-------------------------------------
.TP
.B 0
The mailbox is fully consistent
.\"-------------------------------------
.TP
.B 2
The mailbox is not fully consistent
.\"-------------------------------------
.TP
.B 68
There is not enough information in local metacache to know whether the
mailbox is fully consistent. Use either the
.B \--refresh
parameter or the "full" check.
.\"-------------------------------------
.RE
.PP
.\"-------------------------------------
.SS fts check full
.B doveadm fts check full
[\fB\-u\fP \fIuser\fP|\fB\-A\fP|\fB\-F\fP \fIfile\fP]
[\fB\-S\fP \fIsocket_path\fP]
.RI [ \--print-details ]
.RI [ \--print-mismatches-only ]
.RI [ namespace ]
.PP
This command exists only when the fts_dovecot plugin (Dovecot Pro FTS) is
loaded.
.PP
This command can be used to check FTS indexes for consistency. It performs
a full check to give detailed output of inconsistencies.
.PP
.\"-------------------------------------
.TP
.B \--print-details
Print also IMAP UID numbers and FTS triplet names for each email.
.\"-------------------------------------
.TP
.B \--print-mismatches-only
Print only mailboxes (or emails, with
.B \--print-details
that have inconsistencies.
.\"-------------------------------------
.RE
.PP
Exit codes:
.PP
.\"-------------------------------------
.TP
.B 0
The mailbox is fully consistent
.\"-------------------------------------
.TP
.B 2
The mailbox is not fully consistent
.\"-------------------------------------
.RE
.PP
.\"------------------------------------------------------------------------
@INCLUDE:reporting-bugs@
.\"------------------------------------------------------------------------
.SH SEE ALSO
.BR doveadm (1),
.BR doveadm\-index (1)
.\"-------------------------------------
.PP
Additional resources:
.IP "Full text search indexing"
http://wiki2.dovecot.org/Plugins/FTS

NEWS
v2.3.20 2022-12-12  Aki Tuomi <aki.tuomi@open-xchange.com>

	+ Add dsync_features=no-header-hashes. When this setting is enabled and
	  one dsync side doesn't support mail GUIDs (i.e. imapc), there is no
	  fallback to using header hashes. Instead, dsync assumes that all mails
	  with identical IMAP UIDs contains the same mail contents. This can
	  significantly improve dsync performance with some IMAP servers that
	  don't support caching Date/Message-ID headers.
	+ lua: HTTP client has more settings now, see
	  https://doc.dovecot.org/admin_manual/lua/#dovecot.http.client
	+ replicator: "doveadm replicator status" command now outputs when the
	  next sync is expected for the user.
	- LAYOUT=index: duplicate GUIDs were not cleaned out. Also the list
	  recovery was not optimal.
	- auth: Assert crash would occur when iterating multiple userdb
	  backends.
	- director: Logging into director using master user with
	  auth_master_user_separator character redirected user to a wrong
	  backend, unless master_user_separator setting was also set to the same
	  value. Merged these into auth_master_user_separator.
	- dsync: Couldn't always fix folder GUID conflicts automatically with
	  Maildir format. This resulted in replication repeatedly failing
	  with "Remote lost mailbox GUID".
	- dsync: Failed to migrate INBOX when using namespace prefix=INBOX/,
	  resulting in "Remote lost mailbox GUID" errors.
	- dsync: INBOX was created too early with namespace prefix=INBOX/,
	  resulting a GUID conflict. This may have been resolved automatically,
	  but not always.
	- dsync: v2.3.18 regression: Wrong imapc password with dsync caused
	  Panic: file lib-event.c: line 506 (event_pop_global):
	  assertion failed: (event == current_global_event)
	- imapc: Requesting STATUS for a mailbox with imapc and INDEXPVT
	  configured did not return correct (private) unseen counts.
	- lib-dict: Process would crash when committing data to redis without
	  dict proxy.
	- lib-mail: Corrupted cached BODYSTRUCTURE caused panic during FETCH.
	  Fixes: Panic: file message-part-data.c: line 579 (message_part_is_attachment):
	  assertion failed: (data != NULL). v2.3.13 regression.
	- lib-storage: mail_attribute_dict with dict-sql failed when it tried to
	  lookup empty dict keys.
	- lib: ioloop-kqueue was missing include breaking some BSD builds.
	- oauth2: Using %{oauth2:name} variables could cause useless
	  introspections.
	- pop3: Sending POP3 command with ':' character caused an assert-crash.
	  v2.3.18 regression.
	- replicator: Replication queue had various issues, potentially causing
	  replication requests to become stuck.
	- stats: Invalid Prometheus label names were created with specific
	  histogram group_by configurations. Prometheus rejected these labels.

v2.3.19.1 2022-06-14  Aki Tuomi <aki.tuomi@open-xchange.com>

	- doveadm deduplicate: Non-duplicate mails were deleted.
	  v2.3.19 regression.
	- auth: Crash would occur when iterating multiple backends.
	  Fixes: Panic: file userdb-blocking.c:
	  line 125 (userdb_blocking_iter_next): assertion failed: (ctx->conn != NULL)

v2.3.19 2022-05-10  Aki Tuomi <aki.tuomi@open-xchange.com>

	+ Added mail_user_session_finished event, which is emitted when the mail
	  user session is finished (e.g. imap, pop3, lmtp). It also includes
	  fields with some process statistics information.
	  See https://doc.dovecot.org/admin_manual/list_of_events/ for more
	  information.
	+ Added process_shutdown_filter setting. When an event matches the filter,
	  the process will be shutdown after the current connection(s) have
	  finished. This is intended to reduce memory usage of long-running imap
	  processes that keep a lot of memory allocated instead of freeing it to
	  the OS.
	+ auth: Add cache hit indicator to auth passdb/userdb finished events.
	  See https://doc.dovecot.org/admin_manual/list_of_events/ for more
	  information.
	+ doveadm deduplicate: Performance is improved significantly.
	+ imapc: COPY commands were sent one mail at a time to the remote IMAP
	  server. Now the copying is buffered, so multiple mails can be copied
	  with a single COPY command.
	+ lib-lua: Add a Lua interface to Dovecot's HTTP client library. See
	  https://doc.dovecot.org/admin_manual/lua/ for more information.
	- auth: Cache lookup would use incorrect cache key after username change.
	- auth: Improve handling unexpected LDAP connection errors/hangs.
	  Try to fix up these cases by reconnecting to the LDAP server and
	  aborting LDAP requests earlier.
	- auth: Process crashed if userdb iteration was attempted while auth-workers
	  were already full handling auth requests.
	- auth: db-oauth2: Using %{oauth2:name} variables caused unnecessary
	  introspection requests.
	- dict: Timeouts may have been leaked at deinit.
	- director: Ring may have become unstable if a backend's tag was changed.
	  It could also have caused director process to crash.
	- doveadm kick: Numeric parameter was treated as IP address.
	- doveadm: Proxying can panic when flushing print output. Fixes
	  Panic: file ioloop.c: line 865 (io_loop_destroy): assertion failed:
	  (ioloop == current_ioloop).
	- doveadm sync: BROKENCHAR was wrongly changed to '_' character when
	  migrating mailboxes. This was set by default to %, so any mailbox
	  names containing % characters were modified to "_25".
	- imapc: Copying or moving mails with doveadm to an imapc mailbox could
	  have produced "Error: Syncing mailbox '[...]' failed" Errors. The
	  operation itself succeeded but attempting to sync the destination
	  mailbox failed.
	- imapc: Prevent index log synchronization errors when two or more imapc
	  sessions are adding messages to the same mailbox index files, i.e.
	  INDEX=MEMORY is not used.
	- indexer: Process was slowly leaking memory for each indexing request.
	- lib-fts: fts header filters caused binary content to be sent to the
	  indexer with non-default configuration.

src/plugins/fts/doveadm-fts.c
/* Copyright (c) 2011-2018 Dovecot authors, see the included COPYING file */

#include "lib.h"
#include "str.h"
#include "imap-util.h"
#include "mail-namespace.h"
#include "mail-search.h"
#include "mailbox-list-iter.h"
#include "fts-tokenizer.h"
#include "fts-filter.h"
#include "fts-language.h"
#include "fts-storage.h"
#include "fts-search-args.h"
#include "fts-user.h"
#include "doveadm-print.h"
#include "doveadm-mail.h"
#include "doveadm-mailbox-list-iter.h"
#include "doveadm-fts.h"

const char *doveadm_fts_plugin_version = DOVECOT_ABI_VERSION;

struct fts_tokenize_cmd_context {
	struct doveadm_mail_cmd_context ctx;
	const char *language;
	const char *tokens;
};

struct fts_namespace_cmd_context {
	struct doveadm_mail_cmd_context ctx;
	const char *namespace;
};

static int
cmd_search_box(struct doveadm_mail_cmd_context *ctx,
	       const struct mailbox_info *info)
{
	struct event *event = info->ns->user->event;
	struct mailbox *box;
	struct fts_backend *backend;
	struct fts_result result;
	int ret = 0;

	backend = fts_list_backend(info->ns->list);
	if (backend == NULL) {
		e_error(event, "fts not enabled for %s", info->vname);
		ctx->exit_code = EX_CONFIG;
		return -1;
	}

	i_zero(&result);
	i_array_init(&result.definite_uids, 16);
	i_array_init(&result.maybe_uids, 16);
	i_array_init(&result.scores, 16);

	box = mailbox_alloc(info->ns->list, info->vname, 0);
	if (fts_backend_lookup(backend, box, ctx->search_args->args,
				      FTS_LOOKUP_FLAG_AND_ARGS, &result) < 0) {
		e_error(event, "fts lookup failed");
		doveadm_mail_failed_error(ctx, MAIL_ERROR_TEMP);
		ret = -1;
	} else {
		printf("%s: ", info->vname);
		if (array_count(&result.definite_uids) == 0)
			printf("no results\n");
		else T_BEGIN {
			string_t *str = t_str_new(128);
			imap_write_seq_range(str, &result.definite_uids);
			printf("%s\n", str_c(str));
		} T_END;
		if (array_count(&result.maybe_uids) > 0) T_BEGIN {
			string_t *str = t_str_new(128);
			imap_write_seq_range(str, &result.maybe_uids);
			printf(" - maybe: %s\n", str_c(str));
		} T_END;
		fts_backend_lookup_done(backend);
	}
	mailbox_free(&box);
	array_free(&result.definite_uids);
	array_free(&result.maybe_uids);
	array_free(&result.scores);
	return ret;
}

static int
cmd_fts_lookup_run(struct doveadm_mail_cmd_context *ctx,
		   struct mail_user *user)
{
	const enum mailbox_list_iter_flags iter_flags =
		MAILBOX_LIST_ITER_NO_AUTO_BOXES |
		MAILBOX_LIST_ITER_RETURN_NO_FLAGS;
	struct doveadm_mailbox_list_iter *iter;
	const struct mailbox_info *info;
	int ret = 0;

	iter = doveadm_mailbox_list_iter_init(ctx, user, ctx->search_args,
					      iter_flags);
	while ((info = doveadm_mailbox_list_iter_next(iter)) != NULL) T_BEGIN {
		if (cmd_search_box(ctx, info) < 0)
			ret = -1;
	} T_END;
	if (doveadm_mailbox_list_iter_deinit(&iter) < 0)
		ret = -1;
	return ret;
}

static void
cmd_fts_lookup_init(struct doveadm_mail_cmd_context *_ctx)
{
	struct doveadm_cmd_context *cctx = _ctx->cctx;

	const char *const *query;
	if (!doveadm_cmd_param_array(cctx, "query", &query))
		doveadm_mail_help_name("fts lookup");
	_ctx->search_args = doveadm_mail_build_search_args(query);
}

static struct doveadm_mail_cmd_context *
cmd_fts_lookup_alloc(void)
{
	struct doveadm_mail_cmd_context *ctx;

	ctx = doveadm_mail_cmd_alloc(struct doveadm_mail_cmd_context);
	ctx->v.run = cmd_fts_lookup_run;
	ctx->v.init = cmd_fts_lookup_init;
	return ctx;
}

static int
cmd_fts_expand_run(struct doveadm_mail_cmd_context *ctx,
		   struct mail_user *user)
{
	struct mail_namespace *ns = mail_namespace_find_inbox(user->namespaces);
	struct mailbox *box;
	struct fts_backend *backend;
	string_t *str = t_str_new(128);

	backend = fts_list_backend(ns->list);
	if (backend == NULL) {
		e_error(user->event, "fts not enabled for INBOX");
		ctx->exit_code = EX_CONFIG;
		return -1;
	}

	box = mailbox_alloc(ns->list, "INBOX", 0);
	mail_search_args_init(ctx->search_args, box, FALSE, NULL);

	if (fts_search_args_expand(backend, ctx->search_args) < 0)
		i_fatal("Couldn't expand search args");
	mail_search_args_to_cmdline(str, ctx->search_args->args);
	printf("%s\n", str_c(str));
	mail_search_args_deinit(ctx->search_args);
	mailbox_free(&box);
	return 0;
}

static void
cmd_fts_expand_init(struct doveadm_mail_cmd_context *_ctx)
{
	struct doveadm_cmd_context *cctx = _ctx->cctx;

	const char *const *query;
	if (!doveadm_cmd_param_array(cctx, "query", &query))
		doveadm_mail_help_name("fts expand");
	_ctx->search_args = doveadm_mail_build_search_args(query);
}

static struct doveadm_mail_cmd_context *
cmd_fts_expand_alloc(void)
{
	struct doveadm_mail_cmd_context *ctx;

	ctx = doveadm_mail_cmd_alloc(struct doveadm_mail_cmd_context);
	ctx->v.run = cmd_fts_expand_run;
	ctx->v.init = cmd_fts_expand_init;
	return ctx;
}

static int
cmd_fts_tokenize_run(struct doveadm_mail_cmd_context *_ctx,
		     struct mail_user *user)
{
	struct fts_tokenize_cmd_context *ctx =
		container_of(_ctx, struct fts_tokenize_cmd_context, ctx);

	struct mail_namespace *ns = mail_namespace_find_inbox(user->namespaces);
	struct fts_backend *backend;
	struct fts_user_language *user_lang;
	const struct fts_language *lang = NULL;
	int ret, ret2;
	bool final = FALSE;

	backend = fts_list_backend(ns->list);
	if (backend == NULL) {
		e_error(user->event, "fts not enabled for INBOX");
		_ctx->exit_code = EX_CONFIG;
		return -1;
	}

	if (ctx->language == NULL) {
		struct fts_language_list *lang_list =
			fts_user_get_language_list(user);
		enum fts_language_result result;
		const char *error;

		result = fts_language_detect(lang_list,
		    (const unsigned char *)ctx->tokens, strlen(ctx->tokens),
                    &lang, &error);
		if (lang == NULL)
			lang = fts_language_list_get_first(lang_list);
		switch (result) {
		case FTS_LANGUAGE_RESULT_SHORT:
			e_warning(user->event,
				  "Text too short, can't detect its language - assuming %s",
				  lang->name);
			break;
		case FTS_LANGUAGE_RESULT_UNKNOWN:
			e_warning(user->event,
				  "Can't detect its language - assuming %s",
				  lang->name);
			break;
		case FTS_LANGUAGE_RESULT_OK:
			break;
		case FTS_LANGUAGE_RESULT_ERROR:
			e_error(user->event,
				"Language detection library initialization failed: %s",
				error);
			_ctx->exit_code = EX_CONFIG;
			return -1;
		default:
			i_unreached();
		}
	} else {
		lang = fts_language_find(ctx->language);
		if (lang == NULL) {
			e_error(user->event,
				"Unknown language: %s", ctx->language);
			_ctx->exit_code = EX_USAGE;
			return -1;
		}
	}
	user_lang = fts_user_language_find(user, lang);
	if (user_lang == NULL) {
		e_error(user->event,
			"Language not enabled for user: %s", ctx->language);
		_ctx->exit_code = EX_USAGE;
		return -1;
	}

	fts_tokenizer_reset(user_lang->index_tokenizer);
	for (;;) {
		const char *token, *error;

		if (!final) {
			ret = fts_tokenizer_next(user_lang->index_tokenizer,
				(const unsigned char *)ctx->tokens, strlen(ctx->tokens),
				&token, &error);
		} else {
			ret = fts_tokenizer_final(user_lang->index_tokenizer,
						  &token, &error);
		}
		if (ret < 0)
			break;
		if (ret > 0 && user_lang->filter != NULL) {
			ret2 = fts_filter_filter(user_lang->filter, &token, &error);
			if (ret2 > 0)
				doveadm_print(token);
			else if (ret2 < 0)
				e_error(user->event,
					"Couldn't create indexable tokens: %s",
					error);
		}
		if (ret == 0) {
			if (final)
				break;
			final = TRUE;
		}
	}
	return 0;
}

static void
cmd_fts_tokenize_init(struct doveadm_mail_cmd_context *_ctx)
{
	struct doveadm_cmd_context *cctx = _ctx->cctx;
	struct fts_tokenize_cmd_context *ctx =
		container_of(_ctx, struct fts_tokenize_cmd_context, ctx);

	(void)doveadm_cmd_param_str(cctx, "language", &ctx->language);

	const char *const *args;
	if (!doveadm_cmd_param_array(cctx, "text", &args))
		doveadm_mail_help_name("fts tokenize");

	ctx->tokens = p_strdup(_ctx->pool, t_strarray_join(args, " "));

	doveadm_print_header("token", "token", DOVEADM_PRINT_HEADER_FLAG_HIDE_TITLE);
}

static struct doveadm_mail_cmd_context *
cmd_fts_tokenize_alloc(void)
{
	struct fts_tokenize_cmd_context *ctx;

	ctx = doveadm_mail_cmd_alloc(struct fts_tokenize_cmd_context);
	ctx->ctx.v.run = cmd_fts_tokenize_run;
	ctx->ctx.v.init = cmd_fts_tokenize_init;
	doveadm_print_init(DOVEADM_PRINT_TYPE_FLOW);
	return &ctx->ctx;
}

static int
fts_namespace_find(struct mail_user *user, const char *ns_prefix,
		   struct mail_namespace **ns_r)
{
	struct mail_namespace *ns;

	if (ns_prefix == NULL)
		ns = mail_namespace_find_inbox(user->namespaces);
	else {
		ns = mail_namespace_find_prefix(user->namespaces, ns_prefix);
		if (ns == NULL) {
			e_error(user->event,
				"Namespace prefix not found: %s", ns_prefix);
			return -1;
		}
	}

	if (fts_list_backend(ns->list) == NULL) {
		e_error(user->event, "fts not enabled for user's namespace %s",
			ns_prefix != NULL ? ns_prefix : "INBOX");
		return -1;
	}
	*ns_r = ns;
	return 0;
}

static int
cmd_fts_optimize_run(struct doveadm_mail_cmd_context *_ctx,
		     struct mail_user *user)
{
	struct fts_namespace_cmd_context *ctx =
		container_of(_ctx, struct fts_namespace_cmd_context, ctx);

	const char *ns_prefix = ctx->namespace;
	struct mail_namespace *ns;
	struct fts_backend *backend;

	if (fts_namespace_find(user, ns_prefix, &ns) < 0) {
		doveadm_mail_failed_error(_ctx, MAIL_ERROR_NOTFOUND);
		return -1;
	}
	backend = fts_list_backend(ns->list);
	if (fts_backend_optimize(backend) < 0) {
		e_error(user->event, "fts optimize failed");
		doveadm_mail_failed_error(_ctx, MAIL_ERROR_TEMP);
		return -1;
	}
	return 0;
}

static void
cmd_fts_optimize_init(struct doveadm_mail_cmd_context *_ctx)
{
	struct doveadm_cmd_context *cctx = _ctx->cctx;
	struct fts_namespace_cmd_context *ctx =
		container_of(_ctx, struct fts_namespace_cmd_context, ctx);

	(void)doveadm_cmd_param_str(cctx, "namespace", &ctx->namespace);
}

static struct doveadm_mail_cmd_context *
cmd_fts_optimize_alloc(void)
{
	struct fts_namespace_cmd_context *ctx =
		doveadm_mail_cmd_alloc(struct fts_namespace_cmd_context);
	ctx->ctx.v.run = cmd_fts_optimize_run;
	ctx->ctx.v.init = cmd_fts_optimize_init;
	return &ctx->ctx;
}

static int
cmd_fts_rescan_run(struct doveadm_mail_cmd_context *_ctx, struct mail_user *user)
{
	struct fts_namespace_cmd_context *ctx =
		container_of(_ctx, struct fts_namespace_cmd_context, ctx);


	const char *ns_prefix = ctx->namespace;
	struct mail_namespace *ns;
	struct fts_backend *backend;

	if (fts_namespace_find(user, ns_prefix, &ns) < 0) {
		doveadm_mail_failed_error(_ctx, MAIL_ERROR_NOTFOUND);
		return -1;
	}
	backend = fts_list_backend(ns->list);
	if (fts_backend_rescan(backend) < 0) {
		e_error(user->event, "fts rescan failed");
		doveadm_mail_failed_error(_ctx, MAIL_ERROR_TEMP);
		return -1;
	}
	return 0;
}

static void
cmd_fts_rescan_init(struct doveadm_mail_cmd_context *_ctx)
{
	struct doveadm_cmd_context *cctx = _ctx->cctx;
	struct fts_namespace_cmd_context *ctx =
		container_of(_ctx, struct fts_namespace_cmd_context, ctx);

	(void)doveadm_cmd_param_str(cctx, "namespace", &ctx->namespace);
}

static struct doveadm_mail_cmd_context *
cmd_fts_rescan_alloc(void)
{
	struct fts_namespace_cmd_context *ctx =
		doveadm_mail_cmd_alloc(struct fts_namespace_cmd_context);
	ctx->ctx.v.run = cmd_fts_rescan_run;
	ctx->ctx.v.init = cmd_fts_rescan_init;
	return &ctx->ctx;
}

static struct doveadm_cmd_ver2 fts_commands[] = {
{
	.name = "fts lookup",
	.mail_cmd = cmd_fts_lookup_alloc,
	.usage = DOVEADM_CMD_MAIL_USAGE_PREFIX "<search query>",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_MAIL_COMMON
DOVEADM_CMD_PARAM('\0', "query", CMD_PARAM_ARRAY, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
},
{
	.name = "fts expand",
	.mail_cmd = cmd_fts_expand_alloc,
	.usage = DOVEADM_CMD_MAIL_USAGE_PREFIX "<search query>",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_MAIL_COMMON
DOVEADM_CMD_PARAM('\0', "query", CMD_PARAM_ARRAY, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
},
{
	.name = "fts tokenize",
	.mail_cmd = cmd_fts_tokenize_alloc,
	.usage = DOVEADM_CMD_MAIL_USAGE_PREFIX "<text>",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_MAIL_COMMON
DOVEADM_CMD_PARAM('l', "language", CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAM('\0', "text", CMD_PARAM_ARRAY, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
},
{
	.name = "fts optimize",
	.mail_cmd = cmd_fts_optimize_alloc,
	.usage = DOVEADM_CMD_MAIL_USAGE_PREFIX "[<namespace>]",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_MAIL_COMMON
DOVEADM_CMD_PARAM('\0', "namespace", CMD_PARAM_STR, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
},
{
	.name = "fts rescan",
	.mail_cmd = cmd_fts_rescan_alloc,
	.usage = DOVEADM_CMD_MAIL_USAGE_PREFIX "[<namespace>]",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_MAIL_COMMON
DOVEADM_CMD_PARAM('\0', "namespace", CMD_PARAM_STR, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
},
};

void doveadm_fts_plugin_init(struct module *module ATTR_UNUSED)
{
	unsigned int i;

	for (i = 0; i < N_ELEMENTS(fts_commands); i++)
		doveadm_cmd_register_ver2(&fts_commands[i]);
}

void doveadm_fts_plugin_deinit(void)
{
}
